src/accessories/switchAccessory.ts

```diff
@@ -38,62 +23,50 @@ export class SwitchAccessory extends MiHomePlatformAccessory {
 
     // each service must implement at-minimum the "required characteristics" for the given service type
-    // see https://github.com/homebridge/HAP-NodeJS/blob/master/src/lib/gen/HomeKit.ts
+    // see https://developers.homebridge.io/#/service/Switch
 
     // register handlers for the On/Off Characteristic
     this.service.getCharacteristic(this.platform.Characteristic.On)
-      .on(CharacteristicEventTypes.SET, this.setOn.bind(this))                // SET - bind to the `setOn` method below
-      .on(CharacteristicEventTypes.GET, this.getOn.bind(this));               // GET - bind to the `getOn` method below
+      .on(CharacteristicEventTypes.SET, this.setOn.bind(this))  // SET - bind to the `setOn` method below
+      .on(CharacteristicEventTypes.GET, this.getOn.bind(this)); // GET - bind to the `getOn` method below
+
+    // Update the state of a Characteristic asynchronously instead
+    // of using the `on('get')` handlers.
+    //
+    // Here we update the state.
+    setInterval(() => {
+      // assign the current brightness a random value between 0 and 100
+      const currentBrightness = Math.floor(Math.random() * 100);
+
+      // push the new value to HomeKit
+      this.service.updateCharacteristic(this.platform.Characteristic.Brightness, currentBrightness);
+
+      this.platform.log.debug('Pushed updated current Brightness state to HomeKit:', currentBrightness);
+    }, UPDATE_STATE_INTERVAL);
   }
 
   /**
-   * Handle "SET" requests from HomeKit
-   * These are sent when the user changes the state of an accessory, for example, turning on a Light bulb.
+   * Update the state of a Characteristic asynchronously instead
+   * of using the `on('get')` handlers.
+   * 
+   * Here we update the on / off state using
+   * the `updateCharacteristic` method.
    */
-  setOn(value: CharacteristicValue, callback: CharacteristicSetCallback) {
-    this.states.On = value as boolean;
-
-    this.platform.log.debug('Set Characteristic On ->', value);
-
-    const switchState = this.service.getCharacteristic(this.platform.Characteristic.On);
-
-    if (switchState.value !== value) {
-      const endpoint = this.states.On ? 'power_on' : 'power_off';
-      this.platform.log.debug('start %s', endpoint);
-
-      this.platform.EnergenieApi.toggleSocketPower(this.accessory.context.device.id, endpoint)
-        .then(() => {
-          this.platform.log.debug('%s complete', endpoint);
-          callback(null);
-        })
-        .catch(err => {
-          this.platform.log.debug('Error \'%s\' setting switch state', err);
-          callback(err || new Error('Error setting switch state.'));
-        });
-    } else {
-      callback(null);
-    }
-  }
-
-  /**
-   * Handle the "GET" requests from HomeKit
-   * These are sent when HomeKit wants to know the current state of the accessory.
-  */
-  getOn(callback: CharacteristicGetCallback) {
+  async updateOn() {
```

Comment:
Where do this get called from?

Reply:
Fixed above.

Reply:
This is just the same as `platformAccessory.getOn` and then storing the value - what functionality does this add?

Reply:
There are some differences:
`getOn` calls the callback function.
`updateOn` call the update characteristic function.